clang/lib/Driver/Driver.cpp

```diff
@@ -4709,6 +4733,46 @@ class OffloadingActionBuilder final {
     return C.MakeAction<OffloadAction>(HDep, DDeps);
   }
 
+  // Update Input action to reflect FPGA device archive specifics based
+  // on archive contents.
+  bool updateInputForFPGA(Action *&A, const Arg *InputArg,
+                          DerivedArgList &Args) {
+    std::string InputName = InputArg->getAsString(Args);
+    bool IsFPGAEmulation = C.getDriver().isFPGAEmulationMode();
+    // Only check for FPGA device information when using fpga SubArch.
+    if (A->getType() == types::TY_Object && isObjectFile(InputName))
+      return true;
+
+    auto ArchiveTypeMismatch = [&](bool EmitDiag) {
+      if (EmitDiag)
+        C.getDriver().Diag(clang::diag::warn_drv_mismatch_fpga_archive)
+            << InputName;
+    };
+    // Type FPGA aoco is a special case for static archives
+    if (A->getType() == types::TY_FPGA_AOCO) {
+      if (!hasFPGABinary(C, InputName, types::TY_FPGA_AOCO))
+        return false;
+      A = C.MakeAction<InputAction>(*InputArg, types::TY_FPGA_AOCO);
+      return true;
+    }
+
+    SmallVector<std::pair<types::ID, bool>, 4> FPGAAOCTypes = {
+        {types::TY_FPGA_AOCX, false},
+        {types::TY_FPGA_AOCR, false},
+        {types::TY_FPGA_AOCX_EMU, true},
+        {types::TY_FPGA_AOCR_EMU, true}};
+    for (const auto &ArchiveType : FPGAAOCTypes) {
+      bool binaryFound = hasFPGABinary(C, InputName, ArchiveType.first);
+      if (binaryFound && ArchiveType.second == IsFPGAEmulation) {
```

Comment:
Suggest renaming as per [the guidelines](https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly):
```suggestion
      bool BinaryFound = hasFPGABinary(C, InputName, ArchiveType.first);
      if (BinaryFound && ArchiveType.second == IsFPGAEmulation) {
```